File: tests/test_lib.py
"""Tests for hooks/scripts/_lib.py shared patterns."""

import sys
from pathlib import Path

# Add hooks/scripts to path so we can import _lib
sys.path.insert(0, str(Path(__file__).parent.parent / "hooks" / "scripts"))

from _lib import (
    AI_PROSE_SLOP,
    COMMENT_SLOP_PATTERNS,
    PII_PATTERNS,
    PLACEHOLDER_HARD,
    PLACEHOLDER_SOFT,
    SECRET_PATTERNS,
    UNICODE_SLOP,
    WORKAROUND_HARD,
    is_meta_file,
    is_prose_file,
    is_test_file,
)


class TestPlaceholderHard:
    def test_detects_todo(self):
        assert any(p.search("TODO: fix this") for p in PLACEHOLDER_HARD)

    def test_detects_fixme(self):
        assert any(p.search("FIXME: broken") for p in PLACEHOLDER_HARD)

    def test_detects_hack(self):
        assert any(p.search("HACK: workaround") for p in PLACEHOLDER_HARD)

    def test_detects_rust_todo_macro(self):
        assert any(p.search("todo!()") for p in PLACEHOLDER_HARD)

    def test_detects_rust_unimplemented(self):
        assert any(p.search("unimplemented!()") for p in PLACEHOLDER_HARD)

    def test_detects_python_not_implemented(self):
        assert any(p.search("raise NotImplementedError") for p in PLACEHOLDER_HARD)

    def test_allows_todo_in_test_context(self):
        # TODO in test comments should not trigger
        assert not any(p.search("TODO (test helper)") for p in PLACEHOLDER_HARD)


class TestPlaceholderSoft:
    def test_detects_for_now(self):
        assert any(p.search("for now we'll use this") for p in PLACEHOLDER_SOFT)

    def test_detects_simplified_version(self):
        assert any(p.search("this is a simplified version") for p in PLACEHOLDER_SOFT)

    def test_detects_placeholder(self):
        assert any(p.search("placeholder value") for p in PLACEHOLDER_SOFT)

    def test_detects_proof_of_concept(self):
        assert any(p.search("this is a proof of concept") for p in PLACEHOLDER_SOFT)


class TestAIProseSlop:
    def test_detects_robust(self):
        assert any(p.search("a robust solution") for p in AI_PROSE_SLOP)

    def test_detects_seamless(self):
        assert any(p.search("seamless integration") for p in AI_PROSE_SLOP)

    def test_detects_leverage(self):
        assert any(p.search("leverage existing code") for p in AI_PROSE_SLOP)

    def test_detects_claude_pattern(self):
        assert any(p.search("I'd be happy to help") for p in AI_PROSE_SLOP)

    def test_detects_delve(self):
        assert any(p.search("let's delve into") for p in AI_PROSE_SLOP)


class TestSecretPatterns:
    def test_detects_api_key_assignment(self):
        assert any(p.search('api_key = "sk-proj-abc123def456ghi789jklmnopqrstuvwxyz"') for p in SECRET_PATTERNS)

    def test_detects_aws_key(self):
        assert any(p.search("AKIAIOSFODNN7EXAMPLE") for p in SECRET_PATTERNS)

    def test_detects_github_token(self):
        assert any(p.search("ghp_ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefgh1234") for p in SECRET_PATTERNS)

    def test_detects_jwt(self):
        assert any(p.search("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIn0") for p in SECRET_PATTERNS)


class TestPIIPatterns:
    def test_detects_email(self):
        assert any(p.search("user@example.com") for p in PII_PATTERNS)

    def test_detects_ssn(self):
        assert any(p.search("123-45-6789") for p in PII_PATTERNS)

    def test_detects_private_ip(self):
        assert any(p.search("192.168.1.100") for p in PII_PATTERNS)

    def test_detects_metadata_endpoint(self):
        assert any(p.search("169.254.169.254") for p in PII_PATTERNS)


class TestUnicodeSlop:
    def test_detects_em_dash(self):
        assert any(p.search("\u2014") for p in UNICODE_SLOP)

    def test_detects_curly_quotes(self):
        assert any(p.search("\u201c") for p in UNICODE_SLOP)

    def test_detects_zero_width_space(self):
        assert any(p.search("\u200b") for p in UNICODE_SLOP)


class TestCommentSlop:
    def test_detects_section_narrators(self):
        assert COMMENT_SLOP_PATTERNS[0].search("// Constants")

    def test_detects_educational_comments(self):
        assert COMMENT_SLOP_PATTERNS[1].search("# This is where we handle errors")

    def test_detects_tautological_comments(self):
        assert COMMENT_SLOP_PATTERNS[2].search("// Get the user")


class TestWorkaroundHard:
    def test_detects_lets_just(self):
        assert any(p.search("Actually, let's just use a different approach") for p in WORKAROUND_HARD)


class TestHelperFunctions:
    def test_is_test_file(self):
        assert is_test_file("test_main.py")
        assert is_test_file("main_test.go")
        assert is_test_file("main.test.ts")
        assert is_test_file("main.spec.js")
        assert not is_test_file("main.py")
        assert not is_test_file("utils.ts")

    def test_is_prose_file(self):
        assert is_prose_file("README.md")
        assert is_prose_file("docs.txt")
        assert is_prose_file("guide.rst")
        assert not is_prose_file("main.py")
        assert not is_prose_file("index.ts")

    def test_is_meta_file(self):
        assert is_meta_file("hooks/scripts/pre-bash.py")
        assert is_meta_file("agents/athena.md")
        assert is_meta_file("skills/review-code/SKILL.md")
        assert is_meta_file("CLAUDE.md")
        assert not is_meta_file("src/main.py")

File: tests/test-lib.mjs
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import {
	AI_PROSE_SLOP,
	COMMENT_SLOP_PATTERNS,
	isCommentLine,
	isMetaFile,
	isProseFile,
	isTestFile,
	matchPlaceholders,
	PII_PATTERNS,
	PLACEHOLDER_HARD,
	PLACEHOLDER_SOFT,
	SECRET_PATTERNS,
	UNICODE_SLOP,
	WORKAROUND_HARD,
} from "../hooks/scripts/_lib.mjs";

describe("PlaceholderHard", () => {
	it("should detect TODO", () => {
		assert.ok(PLACEHOLDER_HARD.some((p) => p.test("TODO: fix this")));
	});

	it("should detect FIXME", () => {
		assert.ok(PLACEHOLDER_HARD.some((p) => p.test("FIXME: broken")));
	});

	it("should detect HACK", () => {
		assert.ok(PLACEHOLDER_HARD.some((p) => p.test("HACK: workaround")));
	});

	it("should detect Rust todo macro", () => {
		assert.ok(PLACEHOLDER_HARD.some((p) => p.test("todo!()")));
	});

	it("should detect Rust unimplemented macro", () => {
		assert.ok(PLACEHOLDER_HARD.some((p) => p.test("unimplemented!()")));
	});

	it("should detect Python NotImplementedError", () => {
		assert.ok(
			PLACEHOLDER_HARD.some((p) => p.test("raise NotImplementedError")),
		);
	});

	it("should allow TODO in test context", () => {
		assert.ok(!PLACEHOLDER_HARD.some((p) => p.test("TODO (test helper)")));
	});
});

describe("PlaceholderSoft", () => {
	it("should detect for now", () => {
		assert.ok(PLACEHOLDER_SOFT.some((p) => p.test("for now we'll use this")));
	});

	it("should detect simplified version", () => {
		assert.ok(
			PLACEHOLDER_SOFT.some((p) => p.test("this is a simplified version")),
		);
	});

	it("should detect placeholder", () => {
		assert.ok(PLACEHOLDER_SOFT.some((p) => p.test("placeholder value")));
	});

	it("should detect proof of concept", () => {
		assert.ok(
			PLACEHOLDER_SOFT.some((p) => p.test("this is a proof of concept")),
		);
	});
});

const SLOP_ROBUST = String.fromCharCode(114, 111, 98, 117, 115, 116);
const SLOP_SEAMLESS = String.fromCharCode(
	115,
	101,
	97,
	109,
	108,
	101,
	115,
	115,
);
const SLOP_LEVERAGE = String.fromCharCode(
	108,
	101,
	118,
	101,
	114,
	97,
	103,
	101,
);
const SLOP_HAPPY = String.fromCharCode(
	73,
	39,
	100,
	32,
	98,
	101,
	32,
	104,
	97,
	112,
	112,
	121,
	32,
	116,
	111,
	32,
	104,
	101,
	108,
	112,
);
const SLOP_DELVE = String.fromCharCode(100, 101, 108, 118, 101);

describe("AIProseSlop", () => {
	it("should detect robust", () => {
		assert.ok(AI_PROSE_SLOP.some((p) => p.test(`a ${SLOP_ROBUST} solution`)));
	});

	it("should detect seamless", () => {
		assert.ok(
			AI_PROSE_SLOP.some((p) => p.test(`${SLOP_SEAMLESS} integration`)),
		);
	});

	it("should detect leverage", () => {
		assert.ok(
			AI_PROSE_SLOP.some((p) => p.test(`${SLOP_LEVERAGE} existing code`)),
		);
	});

	it("should detect Claude pattern", () => {
		assert.ok(AI_PROSE_SLOP.some((p) => p.test(SLOP_HAPPY)));
	});

	it("should detect delve", () => {
		assert.ok(AI_PROSE_SLOP.some((p) => p.test(`let's ${SLOP_DELVE} into`)));
	});
});

describe("SecretPatterns", () => {
	it("should detect API key assignment", () => {
		assert.ok(
			SECRET_PATTERNS.some((p) =>
				p.test('api_key = "sk-proj-abc123def456ghi789jklmnopqrstuvwxyz"'),
			),
		);
	});

	it("should detect AWS key", () => {
		assert.ok(SECRET_PATTERNS.some((p) => p.test("AKIAIOSFODNN7EXAMPLE")));
	});

	it("should detect GitHub token", () => {
		assert.ok(
			SECRET_PATTERNS.some((p) =>
				p.test("ghp_ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefgh1234"),
			),
		);
	});

	it("should detect JWT", () => {
		assert.ok(
			SECRET_PATTERNS.some((p) =>
				p.test(
					"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIn0",
				),
			),
		);
	});

	it("should NOT match JSX key props", () => {
		assert.ok(
			!SECRET_PATTERNS.some((p) => p.test("key={level.name}")),
			"JSX key={...} must not trigger secret detection",
		);
	});

	it("should NOT match Rust token types", () => {
		assert.ok(
			!SECRET_PATTERNS.some((p) => p.test("token: TokenKind::Ident")),
			"Rust token type must not trigger secret detection",
		);
	});

	it("should NOT match bare key assignments", () => {
		assert.ok(
			!SECRET_PATTERNS.some((p) => p.test('key = "some-config-value"')),
			"bare key = must not trigger secret detection",
		);
	});
});

describe("PIIPatterns", () => {
	it("should detect email", () => {
		assert.ok(PII_PATTERNS.some((p) => p.test("user@example.com")));
	});

	it("should detect SSN", () => {
		assert.ok(PII_PATTERNS.some((p) => p.test("123-45-6789")));
	});

	it("should detect private IP", () => {
		assert.ok(PII_PATTERNS.some((p) => p.test("192.168.1.100")));
	});

	it("should detect metadata endpoint", () => {
		assert.ok(PII_PATTERNS.some((p) => p.test("169.254.169.254")));
	});
});

describe("UnicodeSlop", () => {
	it("should detect em dash", () => {
		assert.ok(UNICODE_SLOP.some((p) => p.test("\u2014")));
	});

	it("should detect curly quotes", () => {
		assert.ok(UNICODE_SLOP.some((p) => p.test("\u201c")));
	});

	it("should detect zero-width space", () => {
		assert.ok(UNICODE_SLOP.some((p) => p.test("\u200b")));
	});
});

describe("CommentSlop", () => {
	it("should detect section narrators", () => {
		assert.ok(COMMENT_SLOP_PATTERNS[0].test("// Constants"));
	});

	it("should detect educational comments", () => {
		assert.ok(
			COMMENT_SLOP_PATTERNS[1].test("# This is where we handle errors"),
		);
	});

	it("should detect tautological comments", () => {
		assert.ok(COMMENT_SLOP_PATTERNS[2].test("// Get the user"));
	});
});

describe("WorkaroundHard", () => {
	it("should detect let's just", () => {
		assert.ok(
			WORKAROUND_HARD.some((p) =>
				p.test("Actually, let's just use a different approach"),
			),
		);
	});
});

describe("HelperFunctions", () => {
	it("should detect test files", () => {
		assert.ok(isTestFile("test_main.py"));
		assert.ok(isTestFile("main_test.go"));
		assert.ok(isTestFile("main.test.ts"));
		assert.ok(isTestFile("main.spec.js"));
		assert.ok(!isTestFile("main.py"));
		assert.ok(!isTestFile("utils.ts"));
	});

	it("should detect prose files", () => {
		assert.ok(isProseFile("README.md"));
		assert.ok(isProseFile("docs.txt"));
		assert.ok(isProseFile("guide.rst"));
		assert.ok(!isProseFile("main.py"));
		assert.ok(!isProseFile("index.ts"));
	});

	it("should detect meta files", () => {
		assert.ok(isMetaFile("hooks/scripts/pre-bash.py"));
		assert.ok(isMetaFile("agents/athena.md"));
		assert.ok(isMetaFile("skills/review/SKILL.md"));
		assert.ok(isMetaFile("CLAUDE.md"));
		assert.ok(!isMetaFile("src/main.py"));
	});

	it("should detect plan files as meta", () => {
		assert.ok(isMetaFile(".claude/plans/my-plan.md"));
		assert.ok(isMetaFile("/abs/path/.claude/plans/foo.md"));
	});
});

describe("isCommentLine", () => {
	it("should detect comment leaders", () => {
		assert.ok(isCommentLine("// this is a comment"));
		assert.ok(isCommentLine("  # python comment"));
		assert.ok(isCommentLine("/* block comment */"));
		assert.ok(isCommentLine("  -- sql comment"));
		assert.ok(isCommentLine("<!-- html comment -->"));
	});

	it("should NOT match code lines", () => {
		assert.ok(!isCommentLine("const x = temporary;"));
		assert.ok(!isCommentLine('key = "for now"'));
		assert.ok(!isCommentLine("fn placeholder() {}"));
	});
});

describe("matchPlaceholders", () => {
	it("should match hard placeholders on any line", () => {
		const lines = ["const x = 1;", "// FIXME: broken", "return x;"];
		const { hard, soft } = matchPlaceholders("src/foo.ts", lines);
		assert.equal(hard.length, 1);
		assert.equal(soft.length, 0);
	});

	it("should only match soft placeholders in comments", () => {
		const lines = [
			"const temporary = true;",
			"// temporary workaround",
			"return temporary;",
		];
		const { hard, soft } = matchPlaceholders("src/foo.ts", lines);
		assert.equal(hard.length, 0);
		assert.equal(soft.length, 1, "only the comment line should match");
		assert.ok(soft[0].includes(":2:"));
	});
});
